### PR TITLE
Mean apsides and minimal distance

### PR DESCRIPTION
**Pull request by eggrobin**
Add the following to the orbit analysis window:
- a warning if any of the following occur over the course of the mission:
  - dropping below the nominal atmosphere height: reentry,
  - dropping below the highest terrain altitude: collision risk,
  - dropping below the lowest terrain altitude or, on a planet with oceans, sea level, whichever one is higher: collision);
- lowest altitude: the lowest altitude reached over the course of the mission;
- altitude of mean periapsis: the altitude corresponding to the periapsis distance itself corresponding to the mean semimajor axis and mean eccentricity;
- altitude of mean apoapsis: the altitude corresponding to the apoapsis distance itself corresponding to the mean semimajor axis and mean eccentricity;

While the mean periapsis and apoapsis altitudes are not really directly meaningful, since it is customary to describe orbits by their periapsis and apoapsis altitudes, they are a more readily understood description of the size and shape of the orbit.

### CODE REVIEW

File: astronomy/orbital_elements_body.hpp [hunk 1]
```diff
@@ -319,7 +346,9 @@ OrbitalElements::ToClassicalElements(
              : UnwindFrom(classical_elements.back().argument_of_periapsis, ω),
          classical_elements.empty()
              ? Mod(M, 2 * π * Radian)
-             : UnwindFrom(classical_elements.back().mean_anomaly, M)});
+             : UnwindFrom(classical_elements.back().mean_anomaly, M),
+         (1 - e) * equinoctial.a,
+         (1 + e) * equinoctial.a});
```
Comment (pleroy) [inline astronomy/orbital_elements_body.hpp:351]: Could we comment the name of the fields or use designated initializers in this list?  It's impossible to figure out what the various bits mean.

File: ksp_plugin_adapter/orbit_analyser.cs [hunk 2]
```diff
@@ -7,14 +7,21 @@ namespace ksp_plugin_adapter {
 // handling easier, thanks to null-coalescing operators.
 internal static class Formatters {
   // Formats |value| in "N<fractional_digits>" using the Principia culture.
-  public static string FormatN(this double value, int fractional_digits) {
+  public static string FormatN(this double value,
+                               int fractional_digits) {
```
Comment (pleroy) [inline ksp_plugin_adapter/orbit_analyser.cs:11]: Why the change?

File: astronomy/orbital_elements.hpp [hunk 3]
```diff
@@ -138,6 +147,10 @@ class OrbitalElements {
       MassiveBody const& primary,
       Body const& secondary);
 
+  template<typename PrimaryCentred>
+  static std::vector<Length> Distances(
+      DiscreteTrajectory<PrimaryCentred> const& trajectory);
```
Comment (pleroy) [inline astronomy/orbital_elements.hpp:152]: `Distances` is about as uninformative as it gets; make it `DistancesFromPrimary`.

File: astronomy/orbital_elements.hpp [hunk 4]
```diff
@@ -158,19 +171,24 @@ class OrbitalElements {
...
 
-  // |mean_classical_elements_| must have been computed; sets
-  // |mean_*_interval_| accordingly.
-  void ComputeMeanElementIntervals();
+  // |distances_| and |mean_classical_elements_| must have been computed; sets
+  // |distance_interval_| and |mean_*_interval_| accordingly.
+  void ComputeIntervals();
 
   std::vector<EquinoctialElements> osculating_equinoctial_elements_;
+  std::vector<Length> distances_;
```
Comment (pleroy) [inline astronomy/orbital_elements.hpp:179]: `distances_from_primary_`.

File: astronomy/orbital_elements.hpp [hunk 5]
```diff
@@ -158,19 +171,24 @@ class OrbitalElements {
...
   Time sidereal_period_;
   std::vector<EquinoctialElements> mean_equinoctial_elements_;
   std::vector<ClassicalElements> mean_classical_elements_;
   Time anomalistic_period_;
   Time nodal_period_;
   AngularFrequency nodal_precession_;
 
+  Interval<Length> distance_interval_;
```
Comment (pleroy) [inline astronomy/orbital_elements.hpp:187]: `distances_from_primary_interval_`.

File: ksp_plugin_adapter/orbit_analyser.cs [hunk 6]
```diff
@@ -28,9 +35,10 @@ internal static class Formatters {
 
   // Displays an interval of lengths as midpoint±half-width, in km if the
   // half-width is 100 m or more.
```
Comment (pleroy) [inline ksp_plugin_adapter/orbit_analyser.cs:37]: Comment the `offset` parameter.

File: ksp_plugin_adapter/orbit_analyser.cs [hunk 7]
```diff
@@ -236,8 +244,24 @@ internal class OrbitAnalyser : VesselSupervisedWindowRenderer {
...
-      UnityEngine.GUILayout.Label("Orbital elements");
+  private void RenderOrbitalElements(OrbitalElements? elements,
+                                     CelestialBody primary) {
+      double? lowest_distance = elements?.distance.min;
+      LabeledField(
+          "Lowest altitude",
+          (lowest_distance - primary?.Radius)?.FormatAltitude());
+      double? lowest_primary_altitude =
+          primary?.ocean == true ? 0 : primary?.pqsController?.radiusMin;
+      string altitude_warning = lowest_distance < lowest_primary_altitude
```
Comment (pleroy) [inline ksp_plugin_adapter/orbit_analyser.cs:255]: What happens in this comparison if one (or both) of the distances doesn't have a value?

File: serialization/journal.proto [hunk 8]
```diff
@@ -178,6 +178,9 @@ message OrbitalElements {
   required Interval mean_inclination = 7;
   required Interval mean_longitude_of_ascending_nodes = 8;
   required Interval mean_argument_of_periapsis = 9;
+  required Interval mean_periapsis = 10;
```
Comment (pleroy) [inline serialization/journal.proto:181]: Comment before this line:
```
// Added in Gallai.
```

File: ksp_plugin_adapter/orbit_analyser.cs (hunk 7 above)
Comment (eggrobin) [reply ksp_plugin_adapter/orbit_analyser.cs:255]: `false`.

**Comment by eggrobin** [comment]
> The astronomy project already uses C++20, so remove the comments.

`numerics` ends up depending on this (because of the `ToMathematica` overload).

File: astronomy/orbital_elements.hpp (hunk 3 above)
Comment (eggrobin) [reply astronomy/orbital_elements.hpp:152]: `RadialDistances`, and clarified `periapsis_distance` (as opposed to altitude).
